### PR TITLE
feat(component): Add drag and drop support to stateful table component

### LINKED ISSUE #426: Feature Request: Draggable Rows on Stateful Table
**Is your feature request related to a problem? Please describe.**
The current UI for tables allows for draggable rows but the Stateful Table does not support this feature.

**Describe the solution, feature, or improvement you'd like**
I would like to have draggable rows part of Stateful Table like the ones shown here. 
https://www.figma.com/file/Rj63WSTZQc2H4XI9Oi3r5o/MSF-design-(LATEST)?node-id=10702%3A0

Screen Shot 2020-07-29 at 9 35 54 PM

### PR DESCRIPTION
**Pull request by animesh1987**
## What
Add drag and drop support to `StatefulTable` component.

## Why
[Feature request for Drag and Drop](https://github.com/bigcommerce/big-design/issues/426)

Closes #426.

ping @bigcommerce/frontend @deini @chanceaclark

### CODE REVIEW

File: packages/big-design/src/components/StatefulTable/StatefulTable.tsx [hunk 1]
```diff
@@ -86,6 +88,16 @@ const InternalStatefulTable = <T extends TableItem>({
     onSort,
   ]);
 
+  const onDragEnd = useCallback(
+    (from: number, to: number) => {
+      const itemsClone = [...items];
+      const item = itemsClone.splice(from, 1);
+      itemsClone.splice(to, 0, ...item);
```
Comment (chanceaclark) [inline packages/big-design/src/components/StatefulTable/StatefulTable.tsx:95]: 🍹 Could we potentially change this to not rely on mutating an array? Maybe with a combination of `slice` and `spread` operations?

File: packages/docs/PropTables/StatefulTablePropTable.tsx [hunk 2]
```diff
@@ -19,6 +19,11 @@ const statefulTableProps: Prop[] = [
     description: 'The array of items to display.',
     required: true,
   },
+  {
+    name: 'dragAndDrop',
```
Comment (chanceaclark) [inline packages/docs/PropTables/StatefulTablePropTable.tsx:23]: 🍹 Just want to have a discussion about this prop. Should this potentially be something different? Maybe just `draggable`?

**Review by chanceaclark — COMMENTED** [review]
A few things to point out:

- How do know what the new order is when a user has changed it? Seems like we might need a callback.
- Looking at the `Table` documentation, doesn't seem like the new drag and drop props listed. Could you add those in this PR?
- In the **Usage with drag and drop** example in the `Table` documentation, it has some erroneous keys, could we potentially remove `keyField`, `itemName`, and any `isSortable` key/vals to clean up the example?

**Comment by chanceaclark** [comment]
Hey @animesh1987, everything looks great, but could you address these comments?

> * How do know what the new order is when a user has changed it? Seems like we might need a callback.
> * Looking at the `Table` documentation, doesn't seem like the new drag and drop props listed. Could you add those in this PR?
> * In the **Usage with drag and drop** example in the `Table` documentation, it has some erroneous keys, could we potentially remove `keyField`, `itemName`, and any `isSortable` key/vals to clean up the example?

File: packages/big-design/src/components/StatefulTable/StatefulTable.tsx [hunk 3]
```diff
@@ -14,21 +14,36 @@ export interface StatefulTableColumn<T> extends Omit<TableColumn<T>, 'isSortable
 }
 
 export interface StatefulTableProps<T>
-  extends Omit<TableProps<T>, 'columns' | 'pagination' | 'selectable' | 'sortable'> {
+  extends Omit<TableProps<T>, 'columns' | 'pagination' | 'selectable' | 'sortable' | 'onRowDrop'> {
   columns: Array<StatefulTableColumn<T>>;
   pagination?: boolean;
   selectable?: boolean;
   defaultSelected?: T[];
+  onRowDrop?(items: T[]): void;
```
Comment (chanceaclark) [inline packages/big-design/src/components/StatefulTable/StatefulTable.tsx:22]: @deini what do you think of this approach?

Comment (deini) [reply packages/big-design/src/components/StatefulTable/StatefulTable.tsx:22]: The approach is similar to `onSelectionChange`, should be ok 👍

**Review by chanceaclark — APPROVED** [review]
LGTM!

**Pull request merged**